packages/library/src/commonMain/kotlin/io/realm/Realm.kt

```diff
@@ -57,46 +64,72 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
+            updateRealmPointer(nativePointer, versionId)
+            // FIXME What if the result is of a different version than the realm (some other
+            //  write transaction finished before)
+            return result
+        } catch (e: Exception) {
+            throw e
+        }
     }
 
     /**
-     * Modify the underlying Realm file by creating a write transaction on the current thread. Write
-     * transactions automatically commit any changes made when the closure returns unless
-     * [MutableRealm.cancelWrite] was called.
+     * Modify the underlying Realm file while blocking the calling thread until the transaction is
+     * done. Write transactions automatically commit any changes made when the closure returns
+     * unless [MutableRealm.cancelWrite] was called.
      *
      * The write transaction always represent the latest version of data in the Realm file, even if the calling
      * Realm not yet represent this.
      *
      * @param block function that should be run within the context of a write transaction.
      * @return any value returned from the provided write block.
+     *
+     * @throws IllegalStateException if invoked inside an existing transaction.
      */
-    @Suppress("TooGenericExceptionCaught")
     fun <R> writeBlocking(block: MutableRealm.() -> R): R {
-        // While not efficiently to open a new Realm just for a write, it makes it a lot
-        // easier to control the API surface between Realm and MutableRealm
-        val writerRealm = MutableRealm(configuration, dbPointer)
-        try {
-            writerRealm.beginTransaction()
-            val returnValue: R = block(writerRealm)
-            if (writerRealm.commitWrite && !isClosed()) {
-                writerRealm.commitTransaction()
-            }
-            return returnValue
-        } catch (e: Exception) {
-            // Only cancel writes for exceptions. For errors assume that something has gone
-            // horribly wrong and the process is exiting. And canceling the write might just
-            // hide the true underlying error.
-            try {
-                writerRealm.cancelWrite()
-            } catch (cancelException: Throwable) {
-                // Swallow any exception from `cancelWrite` as the primary error is more important.
-                log.error(e)
+        writer.checkInTransaction("Cannot initiate transaction when already in a write transaction")
+        return runBlocking {
+            write(block)
+        }
+    }
+
+    private suspend fun updateRealmPointer(newRealm: NativePointer, newVersion: VersionId) {
+        realmPointerMutex.withLock {
+            log.debug("$version -> $newVersion")
+            if (newVersion >= version) {
+                // FIXME Currently we need this to be a live realm to be able to continue doing
+                //  writeBlocking transactions.
+                dbPointer = RealmInterop.realm_thaw(newRealm)
```

Comment:
This was only to preserve the semantics of emitting live objects to ensure that tests were running as expected until the full frozen architecture is in place and all the tests are reflecting that. It will go in a later iteration.